### PR TITLE
Migrating compound extraction span link tests to weblog

### PR DESCRIPTION
**Pull request by mhlidd**
## Motivation

Currently, we test the results of compound extraction of inconsistent headers through [parametric tests](https://github.com/DataDog/system-tests/blob/main/tests/parametric/test_headers_precedence.py#L653). However, using the parametric app does not truly reflect the behavior exhibited through automatic instrumentation, as we have to manually call extract on the headers passed in, create a span, and add the extracted context/span links to the span. This PR aims to migrate the parametric tests into weblog tests to remove the manual process of imitating this scenario, and have the system-tests represent what actually happens when we have compound extraction on a set of inconsistent headers.

## Changes

Add 2 weblog scenarios that take in `DD_TRACE_PROPAGATION_STYLE_EXTRACT` with values of tracecontext -> datadog -> b3 and datadog -> tracecontext -> b3 respectively. The tests create a span through the `make_distant_call` endpoint and we verify that we get back a span that has the appropriate span links for different scenarios. 

[APMAPI-899](https://datadoghq.atlassian.net/browse/APMAPI-899)
## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

[APMAPI-899]: https://datadoghq.atlassian.net/browse/APMAPI-899?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by mabdinur — APPROVED** [review]
Left some nits overall the approach looks good to me

File: utils/tools.py
Comment (robertomonteromiguel) [inline utils/tools.py:211]: not sure if this is the best place to this method

File: utils/_context/_scenarios/__init__.py
Comment (robertomonteromiguel) [inline utils/_context/_scenarios/__init__.py:108]: you should add the new scenarios here:. github/workflows/run-end-to-end.yml

**Review by robertomonteromiguel — CHANGES_REQUESTED** [review]

File: manifests/cpp.yml
Comment (cbeauchesne) [inline manifests/cpp.yml:243]: Regarding the reason provided, it should be `missing_feature`. See https://github.com/DataDog/system-tests/blob/main/docs/edit/skip-tests.md : 

> * `@irrelevant`: The tested feature/behavior is irrelevant to the library, meaning the feature is either purposefully not supported by the lib or cannot reasonably be implemented
> * `@missing_feature`: The tested feature/behavior does not exist in the library or there is a deficit in the test library that blocks this test from executing for the lib

Same comment for other manifests

File: utils/_context/_scenarios/__init__.py
Comment (cbeauchesne) [inline utils/_context/_scenarios/__init__.py:114]:
```suggestion
        doc="Test compound propagation trace style with w3c headers first",
        scenario_groups=[ScenarioGroup.TRACING_CONFIG, ScenarioGroup.ESSENTIALS],
```

File: utils/_context/_scenarios/__init__.py
Comment (cbeauchesne) [inline utils/_context/_scenarios/__init__.py:123]:
```suggestion
        doc="Test compound propagation trace style with Datadog headers first",
        scenario_groups=[ScenarioGroup.TRACING_CONFIG, ScenarioGroup.ESSENTIALS],
```

File: utils/tools.py
Comment (cbeauchesne) [reply utils/tools.py:211]: Right now, it's only used in `tests/test_distributed.py`, you can place it here as `_retrieve_span_links`. If it'll be used in another file, we'll think where we'llp ut it (very probably in LibraryInterface).

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by cbeauchesne** [comment]
Thank's for tackling this @mhlidd . If you strugle to address any comment, or if you'd like  us to do some of them, please reach us on slack !

File: tests/test_distributed.py
Comment (mabdinur) [inline tests/test_distributed.py:56]:
```suggestion
            and span["trace_id"] == 2  and span["parent_id"] == 10  # Only fetch the trace that is related to the header extractions
```

File: tests/test_distributed.py
Comment (mabdinur) [inline tests/test_distributed.py:76]: Small refactor that could help us avoid needing a second for loop

```suggestion
        assert len(spans) == 1 
        span = spans[0]
        links = retrieve_span_links(span)
        assert len(links) == 2
        link1 = links[0]
        assert link1["trace_id"] == 2
        assert link1["span_id"] == 10
        assert link1["attributes"] == {"reason": "terminated_context", "context_headers": "datadog"}
        assert link1["trace_id_high"] == 2459565876494606882

        link2 = links[1]
        assert link2["trace_id"] == 3
        assert link2["span_id"] == 11744061942159299346
        assert link2["attributes"] == {"reason": "terminated_context", "context_headers": "b3multi"}
        assert link2["trace_id_high"] == 1229782938247303441
```

File: tests/test_distributed.py
Comment (mhlidd) [reply tests/test_distributed.py:56]: This concept would work but value would need to be the tracecontext parent-id. I will add it in my next commit!

File: tests/test_distributed.py
Comment (mabdinur) [inline tests/test_distributed.py:185]: nit: Assertion on tracestate could get tricky. I don't think every library sets this field in the same way.

File: tests/test_distributed.py
Comment (mabdinur) [inline tests/test_distributed.py:269]: nit: can we move  `utils.parametric.spec.trace.retrieve_span_links()` to a new shared module (ex: utils.trace.py) so that we can re-use this function in weblog tests.

**Review by mabdinur — APPROVED** [review]
left two more comments but overall this migration looks good to me. Good job!

File: tests/test_distributed.py
Comment (mhlidd) [reply tests/test_distributed.py:269]: This was addressed by @cbeauchesne [above](https://github.com/DataDog/system-?tests/pull/3499/#discussion_r1856267956). If a new weblog test will use the function, we can move it out then?

File: tests/test_distributed.py
Comment (mhlidd) [reply tests/test_distributed.py:185]: Hmm this test isn't expected to pass as of now, and exists as a result of extracting the tests that are expected to pass versus the ones that test fields that do not have a spec currently. I think keeping it here is fine right now, as it is marked as `missing_feature` due to lack of spec.

**Review by cbeauchesne — CHANGES_REQUESTED** [review]
Some test are failing in the CI. But the rest is fine to me

File: utils/_context/_scenarios/__init__.py
Comment (cbeauchesne) [inline utils/_context/_scenarios/__init__.py:109]: Side question, would it be possible to integrate this conf in an existing scenario, for instance `tracing_config_nondefault_2` and `tracing_config_nondefault_3`.

There are several advantages to use existing scenarios : 

* slower CI time
* higher probability to catch interaction issues
* less work to do to integrate the new scenario on existing CI 

Actually, if it's possible, we should not create individual scenario unless no other scenario is compatible with the conf needed for the new test.

**Comment by cbeauchesne** [comment]
yes it is. Checking

**Comment by cbeauchesne** [comment]
You may have found a real issue in this tracer, the agent does not like either the data from the lib, here is its response : 

`msgp: attempted to decode type \"array\" with method for \"str\" at 21/0\\n`

**Comment by cbeauchesne** [comment]
I've sent a message on slack to the java guild, let's wait for their response

**Comment by cbeauchesne** [comment]
Waiting for #3587 to move forward. Once it's merged, we'll skip the test for spring-boot-native

**Review by cbeauchesne — CHANGES_REQUESTED** [review]
Ok, almost good. There is still one weblog failing : https://github.com/DataDog/system-tests/actions/runs/12075461097/job/33675669767?pr=3499#step:25:57

**Comment by cbeauchesne** [comment]
The java error has been fixed by DataDog/dd-trace-java#8036

**Review by cbeauchesne — APPROVED** [review]
Failures are not related.

**Review by robertomonteromiguel — APPROVED** [review]

**Pull request merged**
